libvast/src/system/node.cpp

```diff
@@ -79,24 +79,55 @@ void peer(node_ptr self, message args) {
   rp.delegate(*peer, peer_atom::value, t, self->state.name);
 }
 
+void collect_component_status(node_ptr self,
+                              caf::response_promise status_promise,
+                              registry& reg) {
+  // Shared state between our response handlers.
+  struct req_state_t {
+    // Keeps track of how many requests are pending.
+    size_t pending = 0;
+    // Promise to the original client request.
+    caf::response_promise rp;
+    // Maps nodes to a map associating components with status information.
+    caf::config_value_map content;
+  };
+  auto req_state = std::make_shared<req_state_t>();
+  req_state->rp = std::move(status_promise);
+  // Pre-fill our result with system stats.
+  auto& sys_stats = req_state->content["system"];
+  auto& sys = self->system();
+  sys_stats.emplace("running-actors", sys.registry().running());
+  sys_stats.emplace("detached-actors", sys.detached_actors());
+  sys_stats.emplace("worker-threads", sys.scheduler().num_workers());
+  // Send out requests and collects ansers.
+  for (auto& [node_name, state_map] : reg.components) {
+    req_state->pending += state_map.size();
+    for (auto& kvp : state_map) {
```

Comment:
I'm  a bit confused about this. It looks like every node maintains a complete view of the topology and each individual component is asked for its status directly? What happens if the registry is out of sync with reality? Would it not make more sense if this request would go to the remote registry and the answers for each node are accumulated locally?

Reply:
The TRACKER maintains the component registry. Here, we ask for that registry to get access to all components. Then, we have to query each component individually with a `status_atom`. If the registry is not up-to-date, then we might get components that have terminated already (this would trigger the error handler below) or miss something that's just coming online.